\section{Our Approach}
In this section we describe the modeling approach we follow.
We begin by discussing the core elements behind Hidden Markov Models.
The discussion is taken from \cite{rabiner2009}

\subsection{HMM development}
\label{hmm_dev}

We are considering two samples for the same individual. One sample is undergoing WGA (sample m605) before sequencing. Sample m585 does not. From the produced sequenced data we then extract the region that corresponds to the chromosome 1. The extracted data is then aligned into non overlapping windows 
of length $L$. Reads of quality less than Q30 were rejected. 



A read is considered only if it satisfies a certain
quality threshold (The \mintinline{c++}{pysam} function \mintinline{c++}{get_query_qualities} which returns query base quality scores at pileup column position). The extracted RD data for each file is then fitted to a normal distribution
  
In our approach we assume that the system can be modelled using four states. Hence, the state set is

\begin{equation}
S = {\text{NORMAL}, \text{TUF}, \text{INSERTION}, \text{DELETION}}
\end{equation}

We can further refine the $S$ set following previous studies e.g.
\cite{coella2007} and \cite{Wang2007} where they assumed that the
$\text{NORMAL}$ state is better described by assuming separate states namely Normal heterozygote and Normal homozygote. 
Furthermore, we assume every state can lead to any other state.
Therefore currently we assume a fully connected model.
The observations set $V$ contains only the RD counts. Since the $\text{TUF}$ state mimics the $\text{DELETION}$ state when considering RD counts (i.e. reduced RD counts) we assume the following:

\begin{itemize}
	\item Regions of low RD observed in both files do not represent $\text{TUF}$ but rather $\text{DELETION}$?
	\item Regions of low RD observed in m605 that are $\text{NORMAL}$ in m585 are indicative of $\text{TUF}$.
\end{itemize}  

.


In our case, we have that the observations consists solely of RD counts. Furthermore, the set of states is the following



We assume the following


We want to fit the an HMM model with $K$ states to the vector of the observed RD counts. We can use the 



There are several items we need to clarify. 

\begin{itemize}
\item How do we initialize the transition probabilities?
\item How do we define the emission probabilities?
\item How do we differentiate between deletion and TUF as the latter mimics deletion in RD?
\end{itemize}


